File: 1901-FindPeakElement2.cpp
class Solution {
public:
    vector<int> findPeakGrid(vector<vector<int>>& mat) {
        int lcol = 0;
        int rcol = mat[0].size() - 1;

        while (lcol <= rcol) {
            int mcol = (lcol + rcol) / 2;
            int mx_row = 0;

            for (int row = 0; row < mat.size(); row++) {
                if (mat[row][mcol] > mat[mx_row][mcol]) {
                    mx_row = row;
                }
            }

            bool leftIsBig = ( mcol - 1 >= 0) && (mat[mx_row][mcol-1] > mat[mx_row][mcol]);
            bool rightIsBig = ( mcol + 1 < mat[0].size()) && (mat[mx_row][mcol+1] > mat[mx_row][mcol]);

            if ((!leftIsBig) && (!rightIsBig)) {
                vector<int> result {mx_row, mcol};
                return result;
            } else if (rightIsBig) {
                lcol = mcol + 1;
            } else {
                rcol = mcol - 1;
            }
        }
        return vector<int> {};
    }
};

File: 1901-FindPeakElement2.py
class Solution:
    def findPeakGrid(self, mat: List[List[int]]) -> List[int]:
        lcol = 0
        rcol = len(mat[0])-1
       
        while lcol <= rcol:
            mx_row = 0
            mcol = (rcol+lcol)//2
            
            for row in range(len(mat)):
                mx_row = row if (mat[row][mcol] >= mat[mx_row][mcol]) else mx_row
            
            leftIsBig    =   mcol-1 >= lcol  and  mat[mx_row][mcol-1] > mat[mx_row][mcol]
            rightIsBig   =   mcol+1 <= rcol    and  mat[mx_row][mcol+1] > mat[mx_row][mcol]
            
            if (not leftIsBig) and (not rightIsBig):
                return [mx_row, mcol]

            elif rightIsBig: 
                lcol = mcol+1 
                
            else:  
                rcol = mcol-1
                
        return []